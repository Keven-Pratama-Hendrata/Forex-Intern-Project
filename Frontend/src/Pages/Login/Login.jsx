import React, { useState } from "react";
import { useNavigate, Link } from "react-router";

import AuthBackground from "../../components/Background/AuthBackground.jsx";
import { Button, LoadingSpinner } from "../../components/common";
import {
  handleChange,
  handleSubmit,
  FormField
} from "./loginUtils.jsx";

const Login = () => {
  const navigate = useNavigate();
  const [loading, setLoading] = useState(false);
  const [form, setForm] = useState({ username: "", password: "" });

  return (
    <AuthBackground circlePosition="login">
      <form
        onSubmit={handleSubmit(form, setLoading, navigate)}
        className="rounded-2xl bg-white/60 p-8 shadow-xl backdrop-blur-md ring-1 ring-white/40"
      >
        <h2 className="mb-1 text-center text-2xl font-extrabold text-gray-800">
          Sign in to Trade FX
        </h2>
        <p className="mb-6 text-center text-sm text-gray-600">
          Ready to conquer the markets?
        </p>

        <FormField
          label="Username"
          type="text"
          name="username"
          placeholder="John Doe"
          value={form.username}
          onChange={handleChange(form, setForm)}
          autoComplete="username"
          required
        />

        <FormField
          label="Password"
          type="password"
          name="password"
          placeholder="••••••••"
          value={form.password}
          onChange={handleChange(form, setForm)}
          autoComplete="current-password"
          required
        />

        <Button type="submit" disabled={loading}>
          {loading ? <LoadingSpinner variant="inline" /> : "Log in"}
        </Button>
      </form>

      <p className="mt-5 text-center text-sm text-gray-700">
        New here?{" "}
        <Link to="/signup" className="font-medium text-sky-950 hover:underline">
          Create an account
        </Link>
      </p>
    </AuthBackground>
  );
};

export default Login;