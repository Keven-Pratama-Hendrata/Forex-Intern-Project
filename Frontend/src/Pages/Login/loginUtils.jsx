import toast from "react-hot-toast";
import { useDispatch } from 'react-redux';
import { loginSuccess } from '../../store/slices/authSlice';
import {
  LoadingSpinner,
  FormField,
  Button,
  handleFormChange,
  validateRequiredFields,
  setUserData
} from "../../components/common";

export const validateForm = (form) => {
  const validation = validateRequiredFields(
    form,
    ['username', 'password'],
    { username: 'Username is required', password: 'Password is required' }
  );

  if (!validation.isValid) {
    toast.error(validation.message);
    return false;
  }
  return true;
};

export const handleLogin = async (form, setLoading, navigate, dispatch) => {
  try {
    setLoading(true);
    const response = await fetch("http://localhost:5001/api/users/login", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(form),
    });

    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.message || "Login failed");
    }

    setUserData(dispatch, loginSuccess, data.token, {
      id: data.userId,
      username: data.userName,
      balances: data.balances
    });

    toast.success("Welcome back ðŸ‘‹");
    navigate('/dashboard');
  } catch (error) {
    toast.error(error.message || "Login failed");
  } finally {
    setLoading(false);
  }
};

export const handleSubmit = (form, setLoading, navigate, dispatch) => async (e) => {
  e.preventDefault();
  if (!validateForm(form)) return;
  await handleLogin(form, setLoading, navigate, dispatch);
};

export { LoadingSpinner, FormField, Button };
export const handleChange = handleFormChange; 