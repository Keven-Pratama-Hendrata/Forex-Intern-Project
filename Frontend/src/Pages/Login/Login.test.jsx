import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { describe, it, expect, jest } from '@jest/globals';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import authReducer from '../../store/slices/authSlice';
import Login from './Login';
import React from 'react';

jest.mock('../../components/Background/AuthBackground.jsx', () => {
  return function MockAuthBackground({ children, circlePosition }) {
    return (
      <div data-testid="auth-background" data-circle-position={circlePosition}>
        {children}
      </div>
    );
  };
});

jest.mock('../../components/common', () => ({
  Button: ({ children, type, disabled, ...props }) => (
    <button
      type={type}
      disabled={disabled}
      data-testid="login-button"
      {...props}
    >
      {children}
    </button>
  ),
  LoadingSpinner: ({ variant }) => (
    <div data-testid="loading-spinner" data-variant={variant}>
      Loading...
    </div>
  ),
  FormField: ({
    label,
    type,
    name,
    placeholder,
    value,
    onChange,
    autoComplete,
    required,
    ...props
  }) => (
    <div data-testid={`form-field-${name}`}>
      <label>{label}</label>
      <input
        type={type}
        name={name}
        placeholder={placeholder}
        value={value}
        onChange={onChange || (() => {})}
        autoComplete={autoComplete}
        required={required}
        data-testid={`input-${name}`}
        {...props}
      />
    </div>
  ),
  handleFormChange: jest.fn(),
  validateRequiredFields: jest.fn(),
  setUserData: jest.fn(),
}));

jest.mock('./loginUtils.jsx', () => {
  const handleChange = jest.fn(() => () => {});
  const handleSubmit = jest.fn(
    (_form, setLoading) =>
      (e) => {
        e?.preventDefault?.();
        setLoading(true);
      },
  );

  const FormField = ({
    label,
    type,
    name,
    placeholder,
    value,
    onChange,
    autoComplete,
    required,
    ...props
  }) => (
    <div data-testid={`form-field-${name}`}>
      <label>{label}</label>
      <input
        type={type}
        name={name}
        placeholder={placeholder}
        value={value}
        onChange={onChange || (() => {})}
        autoComplete={autoComplete}
        required={required}
        data-testid={`input-${name}`}
        {...props}
      />
    </div>
  );

  return { handleChange, handleSubmit, FormField };
});

jest.mock('react-router', () => ({
  useNavigate: () => jest.fn(),
  Link: ({ children, to, ...props }) => (
    <a href={to} {...props}>
      {children}
    </a>
  ),
}));

jest.mock('react-hot-toast', () => ({
  success: jest.fn(),
  error: jest.fn(),
}));

const createMockStore = () =>
  configureStore({
    reducer: { auth: authReducer },
  });

const renderWithProvider = (ui) =>
  render(<Provider store={createMockStore()}>{ui}</Provider>);

describe('Login Component', () => {
  it('renders correctly with default state', () => {
    const { container } = renderWithProvider(<Login />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('toggles to loading state after form submit (covers line 55)', async () => {
    const { container } = renderWithProvider(<Login />);

    const button = screen.getByTestId('login-button');
    expect(button).toHaveTextContent('Log in');
    expect(button).not.toBeDisabled();

    fireEvent.submit(button.closest('form'));

    await waitFor(() =>
      expect(screen.getByTestId('loading-spinner')).toBeInTheDocument(),
    );

    expect(button).toBeDisabled();
    expect(button).toHaveTextContent('Loading...');

    expect(container.firstChild).toMatchSnapshot();
  });
});
