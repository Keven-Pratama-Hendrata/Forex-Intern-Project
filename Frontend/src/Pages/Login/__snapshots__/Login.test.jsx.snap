// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Login Component renders correctly with default state 1`] = `
<div
  data-circle-position="login"
  data-testid="auth-background"
>
  <form
    class="rounded-2xl bg-white/60 p-8 shadow-xl backdrop-blur-md ring-1 ring-white/40"
  >
    <h2
      class="mb-1 text-center text-2xl font-extrabold text-gray-800"
    >
      Sign in to Trade FX
    </h2>
    <p
      class="mb-6 text-center text-sm text-gray-600"
    >
      Ready to conquer the markets?
    </p>
    <div
      data-testid="form-field-username"
    >
      <label>
        Username
      </label>
      <input
        autocomplete="username"
        data-testid="input-username"
        name="username"
        placeholder="John Doe"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      data-testid="form-field-password"
    >
      <label>
        Password
      </label>
      <input
        autocomplete="current-password"
        data-testid="input-password"
        name="password"
        placeholder="••••••••"
        required=""
        type="password"
        value=""
      />
    </div>
    <button
      data-testid="login-button"
      type="submit"
    >
      Log in
    </button>
  </form>
  <p
    class="mt-5 text-center text-sm text-gray-700"
  >
    New here?
     
    <a
      class="font-medium text-sky-950 hover:underline"
      href="/signup"
    >
      Create an account
    </a>
  </p>
</div>
`;

exports[`Login Component toggles to loading state after form submit (covers line 55) 1`] = `
<div
  data-circle-position="login"
  data-testid="auth-background"
>
  <form
    class="rounded-2xl bg-white/60 p-8 shadow-xl backdrop-blur-md ring-1 ring-white/40"
  >
    <h2
      class="mb-1 text-center text-2xl font-extrabold text-gray-800"
    >
      Sign in to Trade FX
    </h2>
    <p
      class="mb-6 text-center text-sm text-gray-600"
    >
      Ready to conquer the markets?
    </p>
    <div
      data-testid="form-field-username"
    >
      <label>
        Username
      </label>
      <input
        autocomplete="username"
        data-testid="input-username"
        name="username"
        placeholder="John Doe"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      data-testid="form-field-password"
    >
      <label>
        Password
      </label>
      <input
        autocomplete="current-password"
        data-testid="input-password"
        name="password"
        placeholder="••••••••"
        required=""
        type="password"
        value=""
      />
    </div>
    <button
      data-testid="login-button"
      disabled=""
      type="submit"
    >
      <div
        data-testid="loading-spinner"
        data-variant="inline"
      >
        Loading...
      </div>
    </button>
  </form>
  <p
    class="mt-5 text-center text-sm text-gray-700"
  >
    New here?
     
    <a
      class="font-medium text-sky-950 hover:underline"
      href="/signup"
    >
      Create an account
    </a>
  </p>
</div>
`;
